### DISCUSSION TITLE
Feature Requests for Tethys Platform 4.0

### DISCUSSION THREAD

**Discussion by swainn** [tethysplatform/tethys#725] [Ideas]
Tethys Platform 4 is around the corner (see below). Please share your ideas on this discussion for what you'd like to see fixed, updated, or added to Tethys Platform for version 4!

Instructions:
1. One idea per comment/post
2. Up-vote the ideas that you like

Tethys Platform Release Schedule:
![image](https://user-images.githubusercontent.com/5123221/132580809-e1800045-fe04-4795-9f20-d3a4174c294c.png)

**Comment by swainn** [comment]
Upgrade to Bootstrap 4/5

**Comment by swainn** [comment]
Upgrade jQuery

**Comment by swainn** [comment]
Integrate JavaScript package managers such as npm or yarn for both the platform and apps.

**Reply by swainn** [reply]
Add npm section to install.yml for apps and define node_modules in public directory.

**Comment by swainn** [comment]
Tethys Layouts

**Comment by swainn** [comment]
Full passthrough wrapper to Django manage.py commands. In other words, `tethys manage` will be able to run any manage.py command and take all arguments.

**Reply by rileyhales** [reply]
cli command to return absolute path to `tethys_portal/manage.py` so its easier to find when you want to launch a tethys portal in debug mode. maybe 
```
tethys manage abspath
tethys manage path
```

**Comment by swainn** [comment]
Django 3.2

**Comment by swainn** [comment]
Schedulers as service settings: #365

**Comment by swainn** [comment]
Update JavaScript APIs to be compatible with module syntax

**Comment by swainn** [comment]
Support Django Forms more natively

**Comment by swainn** [comment]
Upgrade Tethys Docker to use the following docker images:
postgis/postgis
kartoza/geoserver

**Comment by swainn** [comment]
Swap gsconfig for gsconfig-py3 (https://pypi.org/project/gsconfig-py3/).

**Comment by swainn** [comment]
React integration w/ Tethys - create a react Tethys App template

**Reply by swainn** [reply]
Create a Material UI Tethys App scaffold template

**Reply by swainn** [reply]
Create a Bokeh-style app template

**Comment by swainn** [comment]
Create a Material UI Tethys App scaffold template

**Comment by swainn** [comment]
Base template (layout) as an argument for Tethys App scaffolds

**Comment by ezrajrice** [comment]
Add logged timestamp for users who use Token Authentication similar to the "last login" timestamp. Would be useful for apps that provide a service through an API.

**Comment by ganapatinatarajan** [comment]
Is real-time IoT data already a feature of Tethys or could it be included in upcoming version?

**Reply by swainn** [reply]
Hey @GanapatiN Tethys Platform does not have direct support for IoT as of yet, though I'm sure you could integrate it with 3rd party Python libraries. Any sort of IoT devices are you thinking of in particular that would be good for Tethys to support?

**Comment by ganapatinatarajan** [comment]
Sure Nathan. I'll look into that. I was thinking of sensors to measure
basic water quality parameters like TDS, pH or soil moisture and
temperature, for example. I think this functionality would add an extra
dimension to the Tethys platform given the increasing availability and
usage of low-cost sensors for both quantity and quality of environmental
parameters (air, soil, water, etc).
--
Regards,
Ganapati

**Reply by swainn** [reply]
OGC has a standard called Sensor Observation Service (SOS) that may be applicable. 52*North has an implementation: https://52north.org/software/software-projects/sos/
